### ISSUE TITLE
Wrong binding info when forward-referencing

### ISSUE THREAD

**Issue by amasad** [babel/babel#2570]
```
var x = function() {
  if (!y) {
    y = 1;
  }
};
x();
var y = null;
```

The binding object for y:
```
constant: true,
references: 0,
referenced: false,
```

expected:

```
constant: false,
references: 2,
referenced: true,
```

**Comment by amasad** [comment]
I should mention I see this in a `visitor.Scope.enter`.

**Comment by loganfsmyth** [comment]
Duplicate of https://github.com/babel/babel/issues/2550 and fixed in Babel 6 dev branch here: https://github.com/babel/babel/issues/1942

**Comment by amasad** [comment]
@loganfsmyth can we cherrypick to 5.x?

**Comment by loganfsmyth** [comment]
@amasad There are a ton of big changes in 6.x and 5.x is pretty much frozen at this point so that's unlikely. Upgrading to 6.x once it is released should be relatively easy though.

**Comment by sebmck** [comment]
@amasad We can yeah. I'll reopen to do this.

**Comment by sebmck** [comment]
Fixed in 5.8.28.
